x-pack/legacy/plugins/reporting/export_types/common/lib/screenshots/index.ts

```diff
@@ -0,0 +1,152 @@
+    logger.debug(`Creating browser driver factory`);
+
+    const create$ = browserDriverFactory.create({
+      viewport: layout.getBrowserViewport(),
+      browserTimezone,
+    });
+
+    return create$.pipe(
+      mergeMap(({ driver$, exit$, message$, consoleMessage$ }) => {
+        logger.debug('Driver factory created');
+        message$.subscribe((line: string) => {
+          logger.debug(line, ['browser']);
+        });
+        consoleMessage$.subscribe((line: string) => {
+          logger.debug(line, ['browserConsole']);
+        });
+        const screenshot$ = driver$.pipe(
+          mergeMap(
+            (browser: HeadlessBrowser) => openUrl(browser, url, conditionalHeaders, logger),
+            browser => browser
+          ),
+          mergeMap(
+            (browser: HeadlessBrowser) => {
+              logger.debug(
+                'waiting for elements or items count attribute; or not found to interrupt'
+              );
+
+              // the dashboard is using the `itemsCountAttribute` attribute to let us
+              // know how many items to expect since gridster incrementally adds panels
+              // we have to use this hint to wait for all of them
+              const renderSuccess = browser.waitForSelector(
+                `${layout.selectors.renderComplete},[${layout.selectors.itemsCountAttribute}]`
+              );
+              const renderError = checkForToastMessage(browser, layout);
+              return Rx.race(Rx.from(renderSuccess), Rx.from(renderError));
+            },
+            browser => browser
+          ),
+          mergeMap(
+            (browser: HeadlessBrowser) => getNumberOfItems(browser, layout, logger),
+            (browser, itemsCount) => ({ browser, itemsCount })
+          ),
+          mergeMap(
+            async ({ browser, itemsCount }) => {
+              logger.debug('setting viewport');
+              const viewport = layout.getViewport(itemsCount);
+              return await browser.setViewport(viewport);
+            },
+            ({ browser, itemsCount }) => ({ browser, itemsCount })
+          ),
+          mergeMap(
+            ({ browser, itemsCount }) => {
+              logger.debug(`waiting for ${itemsCount} to be in the DOM`);
+              return waitForElementsToBeInDOM(browser, itemsCount, layout);
+            },
+            ({ browser, itemsCount }) => ({ browser, itemsCount })
+          ),
+          mergeMap(
+            ({ browser }) => {
+              // Waiting till _after_ elements have rendered before injecting our CSS
+              // allows for them to be displayed properly in many cases
+              return injectCustomCss(browser, layout, logger);
+            },
+            ({ browser }) => ({ browser })
+          ),
+          mergeMap(
+            async ({ browser }) => {
+              if (layout.positionElements) {
+                logger.debug('positioning elements');
```

Comment:
A lot of these kind of debug log lines have been taken out of the chain and moved to their helper code. I think that's a good way to go, and I think I want to keep continuing it - but in a different PR.

Reply:
Agreed